mne/beamformer/_dics.py

```diff
@@ -1,32 +1,556 @@
 """Dynamic Imaging of Coherent Sources (DICS)."""
 
-# Authors: Roman Goj <roman.goj@gmail.com>
+# Authors: Marijn van Vliet <w.m.vanvliet@gmail.com>
+#          Britta Westner
+#          Susanna Aro <susanna.aro@aalto.fi>
+#          Roman Goj <roman.goj@gmail.com>
 #
 # License: BSD (3-clause)
 
-from copy import deepcopy
+import numbers
 
 import numpy as np
 from scipy import linalg
 
-from ..utils import logger, verbose, warn
+from ..utils import logger, verbose, warn, deprecated
 from ..forward import _subject_from_forward
 from ..minimum_norm.inverse import combine_xyz, _check_reference
 from ..source_estimate import _make_stc
-from ..time_frequency import CrossSpectralDensity, csd_epochs
-from ._lcmv import _prepare_beamformer_input, _setup_picks, _reg_pinv
+from ..time_frequency import csd_epochs
+from ._lcmv import (_prepare_beamformer_input, _setup_picks, _reg_pinv,
+                    _eig_inv, _check_proj_match, _pick_channels_spatial_filter)
 from ..externals import six
 
 
+deprecation_text = ('This function will be removed in 0.17, please use the '
+                    '`make_dics` and `apply_dics_*` functions instead. These '
+                    'functions use a new implementation of DICS that has been '
+                    'more thoroughly tested.')
+
+
 @verbose
-def _apply_dics(data, info, tmin, forward, noise_csd, data_csd, reg,
-                label=None, picks=None, pick_ori=None, real_filter=False,
-                verbose=None):
-    """Dynamic Imaging of Coherent Sources (DICS)."""
+def make_dics(info, forward, csd, reg=0.05, label=None, pick_ori=None,
+              mode='scalar', weight_norm='unit-noise-gain',
+              normalize_fwd=False, real_filter=False, reduce_rank=False,
```

Comment:
I think the switches ``pick_ori`` and ``mode`` are redundant, see ``lcmv``. As I commented somewhere above, ``pick_ori`` in ``lcmv`` controls whether a scalar or vector beamformer is computed.

Reply:
The `mode` parameter determines how `1 / Ck` is computed: either as a 3x3 matrix inversion or as `1/diag(Ck)`. This is not the same as `pick_ori`.

Reply:
I am commenting over here instead of @ https://github.com/mne-tools/mne-python/pull/5066, since the thread is here.
Leaving both choices ``pick_ori`` and ``mode`` is confusing to me. ``pick_ori`` can handle everything related to scalar vs vector beamformer. The choice of ``1 / Ck`` or ``1 / diag(Ck)`` should probably have a separate switch.
The parameters as they are now mean that i could set ``mode="vector"`` and ``pick_ori="max-power"`` which will perform a (post-hoc?) scalar beamformer. I am not sure this is easy on the user ....

Reply:
> The choice of 1 / Ck or 1 / diag(Ck) should probably have a separate switch.

The current `mode` switch is exactly this. I'm also not 100% happy with the name. Any suggestions for a name that is less confusing?

Reply:
Whatever you call it, I would not call one of the options `'vector'` because this is confusing with `pick_ori='vector'`. How about `inversion='matrix' | 'diagonal'`? And if you can be more specific like `noise_inversion` or `norm_inversion` or something, great.

Reply:
+1

Reply:
I can live with `inversion=`